src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -332,77 +333,129 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
+
+                    do
+                    {
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, offset), zero);
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, offset + (nuint)Vector<byte>.Count), zero);
+                        offset += (uint)(2 * Vector<byte>.Count);
+                    } while (offset < stopLoopAtOffset);
+                }
+
+                // At this point, if any data remains to be written, it's strictly less than
+                // 2 * sizeof(Vector) bytes. The loop above had us write an even number of vectors.
+                // If the total byte length instead involves us writing an odd number of vectors, write
+                // one additional vector now. The bit check below tells us if we're in an "odd vector
+                // count" situation.
+
+                if ((totalByteLength & (nuint)Vector<byte>.Count) != 0)
+                {
+                    Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, offset), zero);
+                }
+
+                // It's possible that some small buffer remains to be populated - something that won't
+                // fit an entire vector's worth of data. Instead of falling back to a loop, we'll write
+                // a vector at the very end of the buffer. This may involve overwriting previously
+                // populated data, which is fine since we're just zeroing everything out anyway.
+                // There's no need to perform a length check here because we already performed this
+                // check before entering the vectorized code path.
+
+                Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref refDataAsBytes, totalByteLength - (nuint)Vector<byte>.Count), zero);
             }
             else
             {
-                return; // nothing to write
-            }
+                // If we reached this point, vectorization is disabled, or there are too few
+                // elements for us to vectorize. Fall back to an unrolled loop.
+
+                nuint i = 0;
 
-        Write4To7:
-            Debug.Assert(pointerSizeLength >= 4);
+                // Write 8 elements at a time
+                // Skip this check if "write 8 elements" would've gone down the vectorized code path earlier
 
-            // Write first four and last three.
-            Unsafe.Add(ref ip, 2) = default;
-            Unsafe.Add(ref ip, 3) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -3) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -2) = default;
+                if (!Vector.IsHardwareAccelerated || Vector<byte>.Count / sizeof(IntPtr) > 8) // JIT turns this into constant true or false
+                {
+                    if (pointerSizeLength >= 8)
+                    {
+                        nuint stopLoopAtOffset = pointerSizeLength & ~(nuint)7;
```

Comment:
Although I like this pattern I'm not sure if 
```c#
nuint stopLoopAt = pointerSizeLength - 8;
do
{
    // ...
} while ((i += 8) <= stopLoopAt);
```
is a "bit better".
```diff
-mov       r9,rdx
-and       r9,0FFFFFFFFFFFFFFF8
+lea       r9,[rdx+0FFF8]
```
I doubt it will make a measurable difference in either side.

My thinking is:
The `mov r9,rdx` likely won't reach the cpu's backend, but the `and` needs to be executed.
For the `lea` this can be handled in the AGU.
And it's 3 bytes less code (hooray).
It's a bit more readable, though I assume anyone reading code like this is aware of these tricks.

Out of interest: did you have a reason to choose the bit-hack?

The same applies to other places too (for the record 😉).

Reply:
Bit twiddling hack is muscle memory, since it allows this pattern to work:

```cs
nuint i = 0;
nuint lastOffset = (whatever) & ~(nuint)7;
for (; i != lastOffset; i += 8)
{
    /* do work */
}
```

JIT will use the `and` from the preceding line as the input into the `jcc` at the start of the loop. Allows repurposing the bit-twiddling as the "are you even large enough?" comparison in the first place.

Though to your point, I agree that if we already do the comparison upfront, this twiddling isn't needed. Or we could skip the earlier comparison, eagerly perform the bit-twiddling, and use the result of the twiddling as the comparison.

Reply:
Thanks!